tests/integration/utils.go

```diff
@@ -1011,9 +1013,65 @@ func updateSchema(
 	s *state,
 	action SchemaUpdate,
 ) {
-	_, nodes := getNodesWithIDs(action.NodeID, s.nodes)
-	for _, node := range nodes {
-		results, err := node.AddSchema(s.ctx, action.Schema)
+	// Do some sanitation checks if PolicyIDs are to be substituted, and error out early if invalid usage.
+	if len(action.Replace) > 0 {
+		for substituteLabel := range action.Replace {
+			if substituteLabel == "" {
+				require.Fail(s.t, "Empty substitution label.", s.testCase.Description)
+			}
+
+			howManyLabelsToSub := strings.Count(action.Schema, substituteLabel)
+			if howManyLabelsToSub == 0 {
+				require.Fail(
+					s.t,
+					"Can't do substitution because no label: "+substituteLabel,
+					s.testCase.Description,
+				)
+			}
+		}
+	}
+
+	nodeIDs, nodes := getNodesWithIDs(action.NodeID, s.nodes)
+	for index, node := range nodes {
+		// This schema might be modified if the caller needs some substitution magic done.
+		var modifiedSchema = action.Schema
+
+		// We need to substitute the policyIDs into the `%policyID% place holders.
+		if len(action.Replace) > 0 {
+			nodeID := nodeIDs[index]
+			nodesPolicyIDs := s.policyIDs[nodeID]
+			templateData := map[string]string{}
+			for substituteLabel, replaceWith := range action.Replace {
+				switch repType := replaceWith.(type) {
+				case replacePolicyIndex:
```

Comment:
suggestion: I think the `Replace` interface should be tweaked along with the `replacePolicyIndex` implementation of that interface so that this block is handled by `replacePolicyIndex.Replace`.

By doing it with a switch here we are missing out on a lot of the benefits of using an interface.

Reply:
This is a nice suggestion. I will do.

- [x] remove switch and use the new interface method. Done in last commit.

Reply:
Looks good, thanks Shahzad!  I have a quick follow up suggestion though - `Replacer.Value()` is no longer called and can be removed I think.

Reply:
it is called here:  `policyReplaceIndex, ok := r.Value().(int)`

However can be removed and instead can use `r.value`.

- [x] will do

Reply:
It is calling the concrete `Value()` function there, you can leave `replacePolicyIndex.Value()` there if you like, but there is no reason to keep `Replacer.Value()` atm.

Reply:
Oh yea you were referring about the interface one. Yea I removed it completely, is not needed. Using `r.value` instead